File: solutions/python/grade-school/1/grade_school.py
from bisect import insort


class School:
    def __init__(self):
        self.__roster = []
        self.__added = []

    def add_student(self, name, grade):
        if name not in self.roster():
            insort(self.__roster, (grade, name))
            self.__added.append(True)
        else:
            self.__added.append(False)

    def roster(self):
        return list(name for _, name in self.__roster)

    def grade(self, grade_number):
        return list(name for grade, name in self.__roster if grade == grade_number)

    def added(self):
        return self.__added

File: solutions/cpp/grade-school/3/grade_school.cpp
#include <algorithm>

#include "grade_school.h"

namespace grade_school {
	const std::map<int, std::vector<std::string>>& school::roster() const {
		return roster_;
	}

	void school::add(std::string name, int grade) {
		std::vector<std::string>& student_grade = roster_[grade];
		student_grade.insert(std::upper_bound(student_grade.begin(), student_grade.end(), name), name);
	}

	std::vector<std::string> school::grade(int grade) const {
		auto list = roster_.find(grade);
		if (list == roster_.end())
			return std::vector<std::string>();
		return list->second;
	}
}  // namespace grade_school
